flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitMonitoringFunction.java

```diff
@@ -0,0 +1,234 @@
+import java.util.Set;
+
+/**
+ * This is the single (non-parallel) task which monitors a user-provided path and assigns splits
+ * to downstream tasks for further reading and processing. Which splits will be further processed
+ * depends on the user-provided {@link FileSplitMonitoringFunction.WatchType}.
+ */
+public class FileSplitMonitoringFunction<OUT>
+	extends RichSourceFunction<FileInputSplit> {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(FileSplitMonitoringFunction.class);
+
+	/**
+	 * Specifies when computation will be triggered.
+	 */
+	public enum WatchType {
+		REPROCESS_WITH_APPENDED		// Reprocesses the whole file when new data is appended.
+	}
+
+	/** The path to monitor. */
+	private final String path;
+
+	/** The default parallelism for the job, as this is going to be the parallelism of the downstream readers. */
+	private final int readerParallelism;
+
+	/** The {@link FileInputFormat} to be read. */
+	private FileInputFormat<OUT> format;
+
+	/** How often to monitor the state of the directory for new data. */
+	private final long interval;
+
+	/** Which new data to process (see {@link WatchType}. */
+	private final WatchType watchType;
+
+	private long globalModificationTime;
+
+	private FilePathFilter pathFilter;
+
+	private volatile boolean isRunning = false;
+
+	private Configuration configuration;
+
+	public FileSplitMonitoringFunction(
+		FileInputFormat<OUT> format, String path, Configuration configuration,
+		WatchType watchType, int readerParallelism, long interval) {
+
+		this(format, path, configuration, FilePathFilter.DefaultFilter.getInstance(), watchType, readerParallelism, interval);
+	}
+
+	public FileSplitMonitoringFunction(
+		FileInputFormat<OUT> format, String path, Configuration configuration,
+		FilePathFilter filter, WatchType watchType, int readerParallelism, long interval) {
+
+		this.format = Preconditions.checkNotNull(format);
+		this.path = Preconditions.checkNotNull(path);
+		this.configuration = Preconditions.checkNotNull(configuration);
+
+		Preconditions.checkArgument(interval >= 100,
+			"The specified monitoring interval is smaller than the minimum allowed one (100 ms).");
+		this.interval = interval;
+
+		this.watchType = watchType;
+
+		this.pathFilter = Preconditions.checkNotNull(filter);
+
+		this.readerParallelism = Math.max(readerParallelism, 1);
+		this.globalModificationTime = Long.MIN_VALUE;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void open(Configuration parameters) throws Exception {
+		super.open(parameters);
+		format.configure(this.configuration);
+	}
+
+	/**
+	 * Creates the input splits for the path to be assigned to the downstream tasks.
+	 * Those are going to read their contents for further processing.
+	 */
+	private FileInputSplit[] createInputSplits() throws JobException {
+		FileInputSplit[] inputSplits;
+		try {
+			inputSplits = format.createInputSplits(readerParallelism);
```

Comment:
I think this one line can be moved to `getInputSplitsToForward` which can be renamed to `getInputSplits`. Also, I think we should not use the `JobException`. Just let the `IOExceptions` or whatever bubble through.